### PR TITLE
sql: replace timestamp with cumulative in explain trace

### PR DESCRIPTION
**Pull request by dt**
also change the durstions to always be printed as ms

[Review on Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/4989)

### CODE REVIEW

**Comment by tamird** [comment]
add replace?

**Comment by tamird** [comment]
also durstions

File: sql/trace.go [hunk 1]
```diff
@@ -79,20 +80,35 @@ func (n *explainTraceNode) Next() bool {
 			basePos = n.lastPos + 1
 		}
 
+		// Iterate through once to determine earliest timestamp.
+		var earliest time.Time
+		for _, sp := range n.txn.CollectedSpans {
+			for _, entry := range sp.Logs {
+				if entry.Timestamp.Before(earliest) || earliest.IsZero() {
```
Comment (tamird) [inline sql/trace.go]: nit: seems more natural to check for `earliest.IsZero` first

File: sql/trace.go [hunk 2]
```diff
@@ -79,20 +80,35 @@ func (n *explainTraceNode) Next() bool {
...
+					earliest = entry.Timestamp
+					break
```
Comment (tamird) [inline sql/trace.go]: is this right? i think you need to iterate through them all to find the earliest.

File: sql/trace.go [hunk 3]
```diff
@@ -79,20 +80,35 @@ func (n *explainTraceNode) Next() bool {
...
+				}
+			}
+		}
+
 		for _, sp := range n.txn.CollectedSpans {
 			for i, entry := range sp.Logs {
 				var timeVal string
 				if i > 0 {
-					timeVal = time.Duration(entry.Timestamp.Sub(n.lastTS)).String()
+					timeVal = fmt.Sprintf("%.3fms", time.Duration(entry.Timestamp.Sub(n.lastTS)).Seconds()*1000)
```
Comment (tamird) [inline sql/trace.go]: while you're here, can you call this `duration` and the below `commulativeDuration` for consistency?

File: sql/trace.go [hunk 4]
```diff
@@ -79,20 +80,34 @@ func (n *explainTraceNode) Next() bool {
...
+					parser.DString(sinceStart),
 					parser.DString(timeVal),
 					parser.DInt(basePos + i),
 					parser.DString(sp.Operation),
 					parser.DString(entry.Event),
-				}, vals.AsRow()...))
+				}, vals.AsRow()...)
+
+				// Timestamp is added for sorting, but will be removed after sort.
+				n.rows = append(n.rows, append(cols, parser.DTimestamp{Time: entry.Timestamp}))
```
Comment (tamird) [inline sql/trace.go]: I don't follow -- where's the removal?

**Comment by dt** [comment]
Review status: 0 of 2 files reviewed at latest revision, 4 unresolved discussions.

---

*[sql/trace.go, line 87 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4989#-KCMY1gB15nQWVviIJpR-r1-87:-KCMaCFAcfR9Z_kF9wYI:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/bfa5e2c19763b04da9a4a24f37aab899f1da4316/sql/trace.go#L87)):*
Done.

---

*[sql/trace.go, line 89 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4989#-KCMY1gB15nQWVviIJpR-r1-89:-KCMaCZin75euaQkP6AG:-316571427) ([raw file](https://github.com/cockroachdb/cockroach/blob/bfa5e2c19763b04da9a4a24f37aab899f1da4316/sql/trace.go#L89)):*
in practice the first one in each span is earliest, but it's cheap enough to look at all of them. Done.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4989)*

**Comment by dt** [comment]
Review status: 0 of 2 files reviewed at latest revision, 4 unresolved discussions.

---

*[sql/trace.go, line 98 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4989#-KCMY1gB15nQWVviIJpR-r1-98:-KCMbggjt_mu2Th9dA8s:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/bfa5e2c19763b04da9a4a24f37aab899f1da4316/sql/trace.go#L98)):*
Done.

---

*[sql/trace.go, line 110 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4989#-KCMY1gB15nQWVviIJpR-r2-110:-KCMauJawG6aR0fLleJ5:-1555766035) ([raw file](https://github.com/cockroachdb/cockroach/blob/cda58259b91869f24ab714518e2b29ab01e08a47/sql/trace.go#L110)):*
`sortNode` slices it in `Values()`, just like the other sorts by non-rendered columns.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4989)*

**Comment by tamird** [comment]
:lgtm:

---

Reviewed 1 of 2 files at r1, 1 of 1 files at r3.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4989#-:-KCMbsyuCllB32Ch8xkV:1417088721)*

**Pull request merged**